### PR TITLE
Fix iteration of BTrees.items() in pure-python; and 3.6 support

### LINKED ISSUE #20: Pure-Python BTree.items() can't be iterated, C implementation allows it
(As reported in zopefoundation/zope.proxy#16.)

Given this code:
```python
import BTrees.OOBTree
from zope.security.proxy import Proxy
from zope.security.checker import Checker

bt = BTrees.OOBTree.OOBTree()
bt[''] = 42
items = bt.items()

checker = Checker({})
proxy = Proxy(items, checker)

iter(proxy)
```

It passes using the C implementation of everything (specifically, the C implementation of BTrees), but it fails using pure-python (Python implementation of BTrees):
```
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    iter(proxy)
  File "//zope.filepypy/site-packages/zope/security/proxy.py", line 28, in _wrapper
    checker.check(wrapped, name)
  File "//zope.filepypy/site-packages/zope/security/checker.py", line 229, in check
    raise ForbiddenAttribute(name, object)
ForbiddenAttribute: ('__iter__', <BTrees._base._TreeItems object at 0x000000010da483a0>)
```

(Note that using `CheckerPy` also works with the C implementation of BTrees and fails with the Python version. Also note this happens for both empty trees and for trees that have objects in them. In the empty case, `items()` returns a tuple and the exception is `ForbiddenAttribute: ('__iter__', ())`)

The checker logic doesn't raise an exception if the object doesn't have the `__iter__` attribute:

```C

/*         if name != '__iter__' or hasattr(object, name): */
/*             __traceback_supplement__ = (TracebackSupplement, object) */
/*             raise ForbiddenAttribute, (name, object) */

      if (strcmp("__iter__", MAKE_STRING(name)) == 0
          && ! PyObject_HasAttr(object, name))
        /* We want an attr error if we're asked for __iter__ and we don't
           have it. We'll get one by allowing the access. */
        return 0;
    }
```

The C implementation of `TreeItems` does not have an `__iter__` method:
```
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    print(items.__iter__)
AttributeError: 'OOBTreeItems' object has no attribute '__iter__'
```

Instead, the `TreeItems` implements both `tp_iter` and `tp_iternext` and is its own `iter` (`tp_iter` just returns `self`). But the objects returned by the Python implementation all (necessarily) have an `__iter__`.

Thoughts on how to resolve this? Maybe this isn't a `zope.security` bug, per-se, but whatever sets up the permissions for zope.app should specifically allow `iter` on `BTrees._base._TreeItems`? (That smells bad to me, not least because it fails for the empty/tuple case.)

### PR DESCRIPTION
**Pull request by jamadden**
Also fix ``list(proxy_btree.items())`` (or a list comprehension of the same) in Python 3, which wants the ``__len__`` for a hint.

This is a central place to make sure these all behave consistently.

Fixes #20

### CODE REVIEW

**Comment by jamadden** [comment]
Based on my research in #20, this seemed like a reasonable implementation to fix this. It's centrally located so all users of zope.security will benefit from the fix (which wouldn't happen if we did it in zope.app.security). I think it has to be implemented in Python to be able to handle `__len__` for the C-implemented `XXTreeItems` since they aren't exposed as importable names.

But I'm happy to consider other approaches!

**Comment by jamadden** [comment]
Thank you! A PyPI release when convenient would be helpful.

**Comment by tseaver** [comment]
https://pypi.python.org/pypi/zope.security/4.1.0

**Comment by jamadden** [comment]
Excellent, thank you.

File: src/zope/security/checker.py [hunk 1]
```diff
@@ -770,6 +770,43 @@ def f(): #pragma NO COVER
...
+    # The C implementation of BTree.items() is its own iterator
+    # and doesn't need any special entries to enable iteration.
+    # But the Python implementation has to call __iter__ to be able
+    # to do iteration. Whitelist it so that they behave the same.
+    # In addition, Python 3 will attempt to call __len__ on iterators
+    # for a length hint, so the C implementations also need to be
+    # added to the _iteratorChecker.
+    # We do this here so that all users of zope.security can benefit
+    # without knowing implementation details.
+    # See https://github.com/zopefoundation/zope.security/issues/20
```
Comment (mgedmin) [inline src/zope/security/checker.py]: What about BTree.keys() and .values()?

Comment (jamadden) [inline src/zope/security/checker.py]: Luckily, `.keys()` and `.values()` have the same `XXTreeItems` or `_TreeItems` return value as `.items()`.

Comment (mgedmin) [inline src/zope/security/checker.py]: That is excellent news (and maybe worth mentioning in the comment/added to the test)!

Comment (jamadden) [inline src/zope/security/checker.py]: Good point. #22.

**Pull request merged**